File: backend/app/analyzers/python/duplicates.py
"""
Token-based duplicate detection using the Winnowing algorithm.

Steps:
1. Tokenize each file (Python tokenize module)
2. Compute rolling k-gram hashes over the token stream
3. Apply the Winnowing window to select fingerprints
4. Group files that share fingerprints
5. Report clone groups with file + line-range info
"""
import hashlib
import tokenize
import io
from collections import defaultdict
from pathlib import Path

from app.core.file_walker import FileInfo
from app.models.results import DuplicatesReport, DuplicateBlock, DuplicateInstance

# Tuning parameters
K_GRAM_SIZE = 6       # tokens per k-gram
WINDOW_SIZE = 5       # Winnowing window
MIN_TOKENS = 15       # minimum clone size to report

# Token types to skip (whitespace, comments, encoding markers)
_SKIP_TYPES = {tokenize.COMMENT, tokenize.NL, tokenize.NEWLINE,
               tokenize.INDENT, tokenize.DEDENT, tokenize.ENCODING}


def analyze_duplicates(files: list[FileInfo]) -> DuplicatesReport:
    # Step 1: Tokenize all files
    file_tokens: list[tuple[FileInfo, list[tuple[str, int]]]] = []
    for fi in files:
        try:
            source = fi.path.read_text(encoding="utf-8", errors="ignore")
            tokens = _tokenize(source)
            if tokens:
                file_tokens.append((fi, tokens))
        except Exception:
            continue

    # Step 2: Compute fingerprints per file
    # fingerprint_index maps hash -> list of (FileInfo, line_start, line_end)
    fingerprint_index: dict[int, list[tuple[FileInfo, int, int]]] = defaultdict(list)

    for fi, tokens in file_tokens:
        fps = _winnow(tokens)
        for h, line_start, line_end in fps:
            fingerprint_index[h].append((fi, line_start, line_end))

    # Step 3: Find hashes that appear in >= 2 different files (or same file, diff location)
    groups: list[DuplicateBlock] = []
    seen_block_ids: set[str] = set()
    total_dup_lines = 0

    for h, locations in fingerprint_index.items():
        if len(locations) < 2:
            continue

        # Deduplicate by file+line
        unique = list({(fi.rel_path, ls, le) for fi, ls, le in locations})
        if len(unique) < 2:
            continue

        block_id = hex(h)[2:]
        if block_id in seen_block_ids:
            continue
        seen_block_ids.add(block_id)

        instances: list[DuplicateInstance] = []
        for fi, line_start, line_end in locations:
            snippet = _extract_snippet(fi.path, line_start)
            instances.append(DuplicateInstance(
                file_path=fi.rel_path,
                line_start=line_start,
                line_end=line_end,
                snippet=snippet,
            ))
            total_dup_lines += max(1, line_end - line_start)

        groups.append(DuplicateBlock(
            block_id=block_id,
            instances=instances,
            token_count=K_GRAM_SIZE,
            language="python",
        ))

    return DuplicatesReport(groups=groups, total_duplicate_lines=total_dup_lines)


def _tokenize(source: str) -> list[tuple[str, int]]:
    """Return list of (token_string, line_number), skipping noise tokens."""
    result = []
    try:
        tokens = tokenize.generate_tokens(io.StringIO(source).readline)
        for tok_type, tok_string, (srow, _), _, _ in tokens:
            if tok_type in _SKIP_TYPES:
                continue
            result.append((tok_string, srow))
    except tokenize.TokenError:
        pass
    return result


def _winnow(tokens: list[tuple[str, int]]) -> list[tuple[int, int, int]]:
    """
    Apply Winnowing to produce (hash, line_start, line_end) fingerprints.
    """
    if len(tokens) < K_GRAM_SIZE:
        return []

    strings = [t[0] for t in tokens]
    lines = [t[1] for t in tokens]

    # Build k-gram hashes
    kgram_hashes: list[tuple[int, int, int]] = []
    for i in range(len(strings) - K_GRAM_SIZE + 1):
        gram = "".join(strings[i: i + K_GRAM_SIZE])
        h = _hash(gram)
        kgram_hashes.append((h, lines[i], lines[i + K_GRAM_SIZE - 1]))

    if not kgram_hashes:
        return []

    # Winnowing: slide a window of size WINDOW_SIZE, pick the minimum hash
    fingerprints: list[tuple[int, int, int]] = []
    prev_min_idx = -1

    for i in range(len(kgram_hashes) - WINDOW_SIZE + 1):
        window = kgram_hashes[i: i + WINDOW_SIZE]
        min_h = min(window, key=lambda x: x[0])
        min_idx = i + window.index(min_h)
        if min_idx != prev_min_idx:
            fingerprints.append(min_h)
            prev_min_idx = min_idx

    return fingerprints


def _hash(s: str) -> int:
    return int(hashlib.md5(s.encode()).hexdigest()[:8], 16)


def _extract_snippet(path: Path, line_start: int, max_lines: int = 3) -> str:
    try:
        lines = path.read_text(encoding="utf-8", errors="ignore").splitlines()
        start = max(0, line_start - 1)
        return "\n".join(lines[start: start + max_lines])
    except Exception:
        return ""

File: backend/js_worker/src/duplicates.js
"use strict";

const fs = require("fs");
const acorn = require("acorn");
const crypto = require("crypto");

const K_GRAM_SIZE = 6;
const WINDOW_SIZE = 5;

/**
 * Token-based duplicate detection via Winnowing algorithm.
 */
async function analyzeDuplicates(files) {
  const fingerprints = new Map(); // hash -> [{relPath, lineStart, lineEnd}]

  for (const file of files) {
    let source;
    try {
      source = fs.readFileSync(file.absPath, "utf8");
    } catch {
      continue;
    }

    const tokens = _tokenize(source, file.language);
    if (tokens.length < K_GRAM_SIZE) continue;

    const fps = _winnow(tokens);
    for (const { hash, lineStart, lineEnd } of fps) {
      if (!fingerprints.has(hash)) fingerprints.set(hash, []);
      fingerprints.get(hash).push({ relPath: file.relPath, lineStart, lineEnd });
    }
  }

  const groups = [];
  let totalDupLines = 0;
  const seen = new Set();

  for (const [hash, locations] of fingerprints) {
    if (locations.length < 2) continue;

    // Deduplicate by file+line
    const unique = [...new Map(locations.map((l) => [`${l.relPath}:${l.lineStart}`, l])).values()];
    if (unique.length < 2) continue;

    const blockId = hash.toString(16);
    if (seen.has(blockId)) continue;
    seen.add(blockId);

    const instances = unique.map((loc) => ({
      file_path: loc.relPath,
      line_start: loc.lineStart,
      line_end: loc.lineEnd,
      snippet: _extractSnippet(
        files.find((f) => f.relPath === loc.relPath)?.absPath,
        loc.lineStart,
      ),
    }));

    const dupLines = unique.reduce((sum, l) => sum + Math.max(1, l.lineEnd - l.lineStart), 0);
    totalDupLines += dupLines;

    groups.push({
      block_id: blockId,
      instances,
      token_count: K_GRAM_SIZE,
      language: "javascript",
    });
  }

  return { groups, total_duplicate_lines: totalDupLines };
}

function _tokenize(source, language) {
  // Use acorn token stream for JS; fall back for TS
  const tokens = [];
  try {
    const tokenizer = acorn.tokenizer(source, {
      ecmaVersion: "latest",
      sourceType: "module",
      locations: true,
    });
    for (const tok of tokenizer) {
      if (tok.type.label === "eof") break;
      tokens.push({ value: String(tok.value ?? tok.type.label), line: tok.loc.start.line });
    }
  } catch {
    // Fallback: split by whitespace (less accurate but won't crash on TS syntax)
    const lines = source.split("\n");
    lines.forEach((line, idx) => {
      for (const word of line.trim().split(/\s+/)) {
        if (word) tokens.push({ value: word, line: idx + 1 });
      }
    });
  }
  return tokens;
}

function _winnow(tokens) {
  if (tokens.length < K_GRAM_SIZE) return [];

  // Build k-gram hashes
  const kgrams = [];
  for (let i = 0; i <= tokens.length - K_GRAM_SIZE; i++) {
    const gram = tokens.slice(i, i + K_GRAM_SIZE).map((t) => t.value).join("\x00");
    const hash = _hash(gram);
    kgrams.push({ hash, lineStart: tokens[i].line, lineEnd: tokens[i + K_GRAM_SIZE - 1].line });
  }

  // Winnowing window
  const fps = [];
  let prevMinIdx = -1;
  for (let i = 0; i <= kgrams.length - WINDOW_SIZE; i++) {
    const window = kgrams.slice(i, i + WINDOW_SIZE);
    const minEntry = window.reduce((min, cur) => (cur.hash < min.hash ? cur : min));
    const minIdx = i + window.findIndex((e) => e.hash === minEntry.hash);
    if (minIdx !== prevMinIdx) {
      fps.push(minEntry);
      prevMinIdx = minIdx;
    }
  }

  return fps;
}

function _hash(s) {
  return parseInt(crypto.createHash("md5").update(s).digest("hex").slice(0, 8), 16);
}

function _extractSnippet(absPath, lineStart, maxLines = 3) {
  if (!absPath) return "";
  try {
    const lines = fs.readFileSync(absPath, "utf8").split("\n");
    return lines.slice(Math.max(0, lineStart - 1), lineStart - 1 + maxLines).join("\n");
  } catch {
    return "";
  }
}

module.exports = { analyzeDuplicates };
